# align_it (CSE185 Project Demo)
This is a demonstration project for CSE 185. It is a Python-based tool designed for aligning reads to the Malaria reference genome. This tool is optimized for handling complex genomic sequences with high AT content and repetitive elements, addressing challenges that current alignment tools, optimized for short-read sequencing, may face.
# Installation Instructions
```
git clone https://github.com/cnidyllic/alignit
```
Change into directory and instal tool so it can be used from command line:
```
cd alignit
```
```
python setup.py install
```
Note: if you do not have root access, you can rum the commands above with additional options to install locally: 
```
python setup.py install –user
```
If the install was successful, typing ```align_it –help``` should show a useful message. 

# Basic usage: 
The basic usage of ```align_it``` is:
```
python align_it.py -i <input_FASTQ_file> -r <reference_genome>
```
To run ```align_it``` on a small test example:
Complete usage instructions: 

# align_it options
The required inputs to ```align_it``` is a FASTQ file containing the reads and the reference genome fasta file for alignment. Users may additionally specify the options below:

# File format
The output file format is the same as the bwa mem method, a sam file: See: (link) 

# Contributors
This repository was generated by: , with inspiration from , and work of . 
Please submit a pull request with any corrections and suggestions.

